diff --git a/oggvorbis2fsb5.c b/oggvorbis2fsb5.c
index c290124..681515a 100644
--- a/oggvorbis2fsb5.c
+++ b/oggvorbis2fsb5.c
@@ -4,6 +4,17 @@
 
 // Based on decoder_example.c from the Vorbis library
 
+// Cross platform decorators
+#ifdef _MSC_VER
+#	define TCLE_NATIVE_PACK_BEGIN _Pragma("pack(push, 1)") 
+#	define TCLE_NATIVE_PACK_END _Pragma("pack(pop)")
+#	define TCLE_NATIVE_PACK_ATTRIB
+#else
+#	define TCLE_NATIVE_PACK_BEGIN
+#	define TCLE_NATIVE_PACK_END
+#	define TCLE_NATIVE_PACK_ATTRIB __attribute__((__packed__))
+#endif
+
 #include <math.h>
 #include <stdio.h>
 #include <stdlib.h>
@@ -11,6 +22,9 @@
 #include <string.h>
 #include <vorbis/codec.h>
 
+#include <vector>
+#include <format>
+
 static uint32_t crc32_for_byte(uint32_t r) {
 	for(int j = 0; j < 8; ++j)
 		r = (r & 1? 0: (uint32_t)0xEDB88320L) ^ r >> 1;
@@ -26,7 +40,8 @@ static void crc32(const void *data, size_t n_bytes, uint32_t* crc) {
 		*crc = table[(uint8_t)*crc ^ ((uint8_t*)data)[i]] ^ *crc >> 8;
 }
 
-typedef struct __attribute__((__packed__)) {
+TCLE_NATIVE_PACK_BEGIN;
+typedef struct TCLE_NATIVE_PACK_ATTRIB {
 	uint32_t header;
 	uint32_t version;
 	uint32_t num_samples;
@@ -38,52 +53,137 @@ typedef struct __attribute__((__packed__)) {
 	uint8_t hash[16];
 	uint8_t dummy[8];
 } fsb5_header;
+TCLE_NATIVE_PACK_END
 
-typedef struct __attribute__((__packed__)) {
+TCLE_NATIVE_PACK_BEGIN;
+typedef struct TCLE_NATIVE_PACK_ATTRIB {
 	uint64_t extra_param : 1;
 	uint64_t frequency : 4;
 	uint64_t stereo : 1;
 	uint64_t data_offset : 28;
 	uint64_t samples : 30;
 } fsb5_sample_header;
+TCLE_NATIVE_PACK_END
 
-typedef struct __attribute__((__packed__)) {
+TCLE_NATIVE_PACK_BEGIN;
+typedef struct TCLE_NATIVE_PACK_ATTRIB {
 	uint32_t has_next : 1;
 	uint32_t size : 24;
 	uint32_t chunk_type : 7;
 } fsb5_sample_extra_header;
+TCLE_NATIVE_PACK_END
 
-typedef struct __attribute__((__packed__)) {
+TCLE_NATIVE_PACK_BEGIN;
+typedef struct TCLE_NATIVE_PACK_ATTRIB {
 	uint8_t channels;
 } fsb5_sample_extra_channels_header;
+TCLE_NATIVE_PACK_END
 
-typedef struct __attribute__((__packed__)) {
+TCLE_NATIVE_PACK_BEGIN;
+typedef struct TCLE_NATIVE_PACK_ATTRIB {
 	uint32_t frequency;
 } fsb5_sample_extra_frequency_header;
+TCLE_NATIVE_PACK_END
 
-typedef struct __attribute__((__packed__)) {
+TCLE_NATIVE_PACK_BEGIN;
+typedef struct TCLE_NATIVE_PACK_ATTRIB {
 	uint32_t loop_start;
 	uint32_t loop_end;
 } fsb5_sample_extra_loop_header;
+TCLE_NATIVE_PACK_END
 
-typedef struct __attribute__((__packed__)) {
+TCLE_NATIVE_PACK_BEGIN;
+typedef struct TCLE_NATIVE_PACK_ATTRIB {
 	uint32_t crc32;
 	uint32_t position_offset_table_length;
 } fsb5_sample_extra_vorbis_header;
+TCLE_NATIVE_PACK_END
 
-typedef struct __attribute__((__packed__)) {
+TCLE_NATIVE_PACK_BEGIN;
+typedef struct TCLE_NATIVE_PACK_ATTRIB {
 	uint32_t granulepos;
 	uint32_t offset;
 } fsb5_sample_extra_vorbis_entry_header;
+TCLE_NATIVE_PACK_END
 
-typedef struct __attribute__((__packed__)) linked_list_for_ogg_packets_struct {
+TCLE_NATIVE_PACK_BEGIN;
+typedef struct TCLE_NATIVE_PACK_ATTRIB linked_list_for_ogg_packets_struct {
 	unsigned char *packet;
 	long bytes;
 	ogg_int64_t granulepos;
 	struct linked_list_for_ogg_packets_struct *next;
 } linked_list_for_ogg_packets;
+TCLE_NATIVE_PACK_END
+
+// The entire file is already loaded into memory,
+// This is a simple wrapper to mimick the behavior of fread to avoid code restucturing
+typedef struct {
+	unsigned char const* data;
+	size_t size;
+	size_t pos;
+} MemFile;
+
+size_t mem_fread(void* buffer, size_t size, size_t count, MemFile& mf)
+{
+	size_t total = size * count;
+	size_t remaining = mf.size - mf.pos;
+
+	if (remaining == 0)
+		return 0;
+
+	if (total > remaining)
+		total = remaining;
+
+	memcpy(buffer, mf.data + mf.pos, total);
+	mf.pos += total;
+
+	return total / size; // just like fread: returns number of *items* read
+}
+
+size_t mem_fwrite(const void* buffer, size_t size, size_t count, std::vector<unsigned char>& out)
+{
+	size_t total = size * count;
+	const unsigned char* p = static_cast<const unsigned char*>(buffer);
+
+	// Append to the vector
+	out.insert(out.end(), p, p + total);
+
+	return count; // same return semantics as fwrite: number of *items* written
+}
+
+#include <string>
+
+static std::string kErrorLog;
+static std::string kInfoLog;
+
+extern "C" __declspec(dllexport) void ns_anthofoxo_tcle_get_info_log(void const** pData, int* size) {
+	*pData = kErrorLog.data();
+	*size = kErrorLog.size();
+}
+
+extern "C" __declspec(dllexport) void ns_anthofoxo_tcle_get_error_log(void const** pData, int* size) {
+	*pData = kInfoLog.data();
+	*size = kInfoLog.size();
+}
+
+
+extern "C" __declspec(dllexport) void ns_anthofoxo_tcle_free(void* ptr) {
+	free(ptr);
+}
+
+// long name to guarentee no name collision
+extern "C" __declspec(dllexport) int ns_anthofoxo_tcle_ogg_vorbis_fsb5_convert(void* inData, int inSize, void** outData, int* outSize) {
+	kErrorLog.clear();
+	kInfoLog.clear();
+
+	// construct our memory view
+	MemFile infile;
+	infile.data = (unsigned char*)inData;
+	infile.size = inSize;
+	infile.pos = 0;
+
+	// vvv Original code vvv
 
-int main(int argc, char** argv) {
 	ogg_sync_state oy;   /* sync and verify incoming physical bitstream */
 	ogg_stream_state os; /* take physical pages, weld into a logical stream of packets */
 	ogg_page og;         /* one Ogg bitstream page. Vorbis packets are inside */
@@ -106,13 +206,13 @@ int main(int argc, char** argv) {
 	fsb5_sample_extra_loop_header fselh;
 	fsb5_sample_extra_header fseh_vorbisdata;
 	fsb5_sample_extra_vorbis_header fsevh;
 	fsb5_sample_extra_vorbis_entry_header *fseveh;
 
 	linked_list_for_ogg_packets *ll;
 	linked_list_for_ogg_packets *ll_last;
 	linked_list_for_ogg_packets *ll_current;
 
 	char *buffer;
 	int bytes;
 
 
@@ -125,18 +225,6 @@ int main(int argc, char** argv) {
 
 	ogg_sync_init(&oy); /* Now we can read pages */
 
-	if (argc < 2)
-	{
-		fprintf(stderr, "Error: not enough arguments for input file\n");
-		return 2;
-	}
-	FILE *infile = fopen(argv[1], "rb");
-	if (!infile)
-	{
-		fprintf(stderr, "Error: could not open input file\n");
-		return 1;
-	}
-
 	while (1) { /* we repeat if the bitstream is chained */
 		int eos = 0;
 		int i;
@@ -146,9 +234,9 @@ int main(int argc, char** argv) {
 			 stream initial header) We need the first page to get the stream
 			 serialno. */
 
 		/* submit a 4k block to libvorbis' Ogg layer */
 		buffer = ogg_sync_buffer(&oy, 4096);
-		bytes = fread(buffer, 1, 4096, infile);
+		bytes = mem_fread(buffer, 1, 4096, infile);
 		ogg_sync_wrote(&oy, bytes);
 
 		/* Get the first page. */
@@ -158,8 +246,8 @@ int main(int argc, char** argv) {
 				break;
 
 			/* error case.  Must not be Vorbis data */
-			fprintf(stderr, "Input does not appear to be an Ogg bitstream.\n");
-			return 1;
+			kErrorLog += "Input does not appear to be an Ogg bitstream.\n";
+			return false;
 		}
 
 		/* Get the serial number and set up the rest of decode. */
@@ -177,31 +265,32 @@ int main(int argc, char** argv) {
 		vorbis_comment_init(&vc);
 		if (ogg_stream_pagein(&os, &og) < 0) {
 			/* error; stream version mismatch perhaps */
-			fprintf(stderr, "Error reading first page of Ogg bitstream data.\n");
-			return 1;
+			kErrorLog += "Error reading first page of Ogg bitstream data.\n";
+			return false;
 		}
 
 		if (ogg_stream_packetout(&os, &op) != 1) {
 			/* no page? must not be vorbis */
-			fprintf(stderr, "Error reading initial header packet.\n");
-			return 1;
+			kErrorLog += "Error reading initial header packet.\n";
+			return false;
 		}
 
 		if (vorbis_synthesis_headerin(&vi, &vc, &op) < 0) {
 			/* error case; not a vorbis header */
-			fprintf(stderr, "This Ogg bitstream does not contain Vorbis audio data.\n");
-			return 1;
+
+			kErrorLog += "This Ogg bitstream does not contain Vorbis audio data.\n";
+			return false;
 		}
 
 		/* At this point, we're sure we're Vorbis. We've set up the logical
 			 (Ogg) bitstream decoder. Get the comment and codebook headers and
 			 set up the Vorbis decoder */
 
 		/* The next two packets in order are the comment and codebook headers.
 			 They're likely large and may span multiple pages. Thus we read
 			 and submit data until we get our two packets, watching that no
 			 pages are missing. If a page is missing, error out; losing a
 			 header page is the only place where missing data is fatal. */
 
 		i = 0;
 		while (i < 2) {
@@ -217,14 +306,13 @@ int main(int argc, char** argv) {
 						if (result == 0)
 							break;
 						if (result < 0) {
-							/* Uh oh; data at some point was corrupted or missing! We can't tolerate that in a header.  Die. */
-							fprintf(stderr, "Corrupt secondary header.  Exiting.\n");
-							return 1;
+							kErrorLog += "Corrupt secondary header.  Exiting.\n";
+							return false;
 						}
 						result = vorbis_synthesis_headerin(&vi, &vc, &op);
 						if (result < 0) {
-							fprintf(stderr, "Corrupt secondary header.  Exiting.\n");
-							return 1;
+							kErrorLog += "Corrupt secondary header.  Exiting.\n";
+							return false;
 						}
 						i++;
 						if (i == 2)
@@ -237,24 +325,23 @@ int main(int argc, char** argv) {
 			}
 			/* no harm in not checking before adding more */
 			buffer = ogg_sync_buffer(&oy, 4096);
-			bytes = fread(buffer, 1, 4096, infile);
+			bytes = mem_fread(buffer, 1, 4096, infile);
 			if (bytes == 0 && i < 2) {
-				fprintf(stderr, "End of file before finding all Vorbis headers!\n");
-				return 1;
+				kErrorLog += "End of file before finding all Vorbis headers!\n";
+				return false;
 			}
 			ogg_sync_wrote(&oy, bytes);
 		}
 
 		/* Throw the comments plus a few lines about the bitstream we're decoding */
 		{
 			char **ptr = vc.user_comments;
 			while (*ptr) {
-				fprintf(stderr, "%s\n", *ptr);
+				kInfoLog += std::format("{}\n", *ptr);
 				++ptr;
 			}
-			fprintf(stderr, "\nBitstream is %d channel, %ldHz\n", vi.channels,
-							vi.rate);
-			fprintf(stderr, "Encoded by: %s\n\n", vc.vendor);
+			kInfoLog += std::format("\nBitstream is {} channel, {}Hz\n", vi.channels, vi.rate);
+			kInfoLog += std::format("Encoded by: {}\n\n", vc.vendor);
 		}
 
 		/* OK, got and parsed all three headers. Initialize the Vorbis packet->PCM decoder. */
@@ -270,8 +357,9 @@ int main(int argc, char** argv) {
 					if (result == 0)
 						break;          /* need more data */
 					if (result < 0) { /* missing or corrupt data at this page position */
-						fprintf(stderr, "Corrupt or missing data in bitstream; continuing...\n");
-					} else {
+						kErrorLog += "Corrupt or missing data in bitstream; continuing...\n";
+					}
+					else {
 						ogg_stream_pagein(&os, &og); /* can safely ignore errors at this point */
 						while (1) {
 							result = ogg_stream_packetout(&os, &op);
@@ -280,11 +368,12 @@ int main(int argc, char** argv) {
 								break; /* need more data */
 							if (result < 0) { /* missing or corrupt data at this page position */
 								/* no reason to complain; already complained above */
-							} else {
+							}
+							else {
 								/* we have a packet.  Decode it */
-								ll_current = malloc(sizeof(linked_list_for_ogg_packets));
+								ll_current = (decltype(ll_current))malloc(sizeof(linked_list_for_ogg_packets));
 								memset(ll_current, 0, sizeof(linked_list_for_ogg_packets));
-								ll_current->packet = malloc(op.bytes);
+								ll_current->packet = (decltype(ll_current->packet))malloc(op.bytes);
 								memcpy(ll_current->packet, op.packet, op.bytes);
 								ll_current->bytes = op.bytes;
 								ll_current->granulepos = op.granulepos;
@@ -305,7 +394,7 @@ int main(int argc, char** argv) {
 				}
 				if (!eos) {
 					buffer = ogg_sync_buffer(&oy, 4096);
-					bytes = fread(buffer, 1, 4096, infile);
+					bytes = mem_fread(buffer, 1, 4096, infile);
 					ogg_sync_wrote(&oy, bytes);
 					if (bytes == 0)
 						eos = 1;
@@ -316,8 +405,9 @@ int main(int argc, char** argv) {
 
 			vorbis_block_clear(&vb);
 			vorbis_dsp_clear(&vd);
-		} else {
-			fprintf(stderr, "Error: Corrupt header during playback initialization.\n");
+		}
+		else {
+			kErrorLog += "Error: Corrupt header during playback initialization.\n";
 		}
 
 		/* clean up this logical bitstream; before exit we see if we're followed by another [chained] */
@@ -333,17 +423,10 @@ int main(int argc, char** argv) {
 	/* OK, clean up the framer */
 	ogg_sync_clear(&oy);
 
-	if (argc < 3)
-	{
-		fprintf(stderr, "Error: not enough arguments for output file\n");
-		return 2;
-	}
-	FILE *outfile = fopen(argv[2], "wb");
-	if (!outfile)
-	{
-		fprintf(stderr, "Error: could not open output file\n");
-		return 1;
-	}
+	// mimick outfile from fopen
+	std::vector<unsigned char> outfile;
+	//
+	
 	int fsb_header_size = 0;
 	int fsb_sample_header_size = 0;
 	int packet_count = 0;
@@ -373,74 +456,74 @@ int main(int argc, char** argv) {
 	fsh.extra_param = 1;
 	switch (vi.rate)
 	{
 		case 48000:
 		{
 			fsh.frequency = 9;
 			break;
 		}
 		case 44100:
 		{
 			fsh.frequency = 8;
 			break;
 		}
 		case 32000:
 		{
 			fsh.frequency = 7;
 			break;
 		}
 		case 24000:
 		{
 			fsh.frequency = 6;
 			break;
 		}
 		case 22050:
 		{
 			fsh.frequency = 5;
 			break;
 		}
 		case 16000:
 		{
 			fsh.frequency = 4;
 			break;
 		}
 		case 11025:
 		{
 			fsh.frequency = 3;
 			break;
 		}
 		case 11000:
 		{
 			fsh.frequency = 2;
 			break;
 		}
 		case 8000:
 		{
 			fsh.frequency = 1;
 			break;
 		}
 		default:
 		{
 			fsh.frequency = 0;
 			break;
 		}
 	}
 	switch (vi.channels)
 	{
		case 2:
		{
			fsh.stereo = 1;
			break;
		}
		case 1:
		{
			fsh.stereo = 0;
			break;
		}
		default:
		{
			fsh.stereo = 0;
			break;
		}
 	}
 	fsh.samples = 0;
 	fsb_sample_header_size += sizeof(fsh);
@@ -463,9 +546,10 @@ int main(int argc, char** argv) {
 		fsb_sample_header_size += sizeof(fseh_frequency) + fseh_frequency.size;
 	}
 	memset(&fseh_loop, 0, sizeof(fseh_loop));
+#if 0 // our use case never needs to account for this, simply exclude it from complilation
 	if (argc >= 5)
 	{
 		char *end;
 		fseh_loop.has_next = 1;
 		fseh_loop.size = sizeof(fselh);
 		fseh_loop.chunk_type = 3; // loop
@@ -473,6 +557,7 @@ int main(int argc, char** argv) {
 		fselh.loop_end = (uint32_t)strtoul(argv[4], &end, 10);
 		fsb_sample_header_size += sizeof(fseh_loop) + fseh_loop.size;
 	}
+#endif
 	memset(&fseh_vorbisdata, 0, sizeof(fseh_vorbisdata));
 	fseh_vorbisdata.has_next = 0;
 	fseh_vorbisdata.size = sizeof(fsevh) + sizeof(fsb5_sample_extra_vorbis_entry_header) * packet_count;
@@ -484,8 +569,8 @@ int main(int argc, char** argv) {
 	fsb_sample_header_size += data_padding;
 	fh.sample_header_size = fsb_sample_header_size;
 	fsb_header_size += fsb_sample_header_size;
-	
-	fseveh = malloc(sizeof(fsb5_sample_extra_vorbis_entry_header) * packet_count);
+
+	fseveh = (decltype(fseveh))malloc(sizeof(fsb5_sample_extra_vorbis_entry_header) * packet_count);
 	packet_count = 0;
 	ll_current = ll;
 	int total_offset = 0;
@@ -507,40 +592,46 @@ int main(int argc, char** argv) {
 		ll_current = ll_current->next;
 	}
 	fsh.samples = max_granulepos;
-	fwrite(&fh, 1, sizeof(fh), outfile);
-	fwrite(&fsh, 1, sizeof(fsh), outfile);
+	mem_fwrite(&fh, 1, sizeof(fh), outfile);
+	mem_fwrite(&fsh, 1, sizeof(fsh), outfile);
 	if (fseh_channels.has_next)
 	{
-		fwrite(&fseh_channels, 1, sizeof(fseh_channels), outfile);
-		fwrite(&fsech, 1, sizeof(fsech), outfile);
+		mem_fwrite(&fseh_channels, 1, sizeof(fseh_channels), outfile);
+		mem_fwrite(&fsech, 1, sizeof(fsech), outfile);
 	}
 	if (fseh_frequency.has_next)
 	{
-		fwrite(&fseh_frequency, 1, sizeof(fseh_frequency), outfile);
-		fwrite(&fsefh, 1, sizeof(fsefh), outfile);
+		mem_fwrite(&fseh_frequency, 1, sizeof(fseh_frequency), outfile);
+		mem_fwrite(&fsefh, 1, sizeof(fsefh), outfile);
 	}
 	if (fseh_loop.has_next)
 	{
-		fwrite(&fseh_loop, 1, sizeof(fseh_loop), outfile);
-		fwrite(&fselh, 1, sizeof(fselh), outfile);
+		mem_fwrite(&fseh_loop, 1, sizeof(fseh_loop), outfile);
+		mem_fwrite(&fselh, 1, sizeof(fselh), outfile);
 	}
-	fwrite(&fseh_vorbisdata, 1, sizeof(fseh_vorbisdata), outfile);
-	fwrite(&fsevh, 1, sizeof(fsevh), outfile);
-	fwrite(fseveh, 1, sizeof(fsb5_sample_extra_vorbis_entry_header) * packet_count, outfile);
+	mem_fwrite(&fseh_vorbisdata, 1, sizeof(fseh_vorbisdata), outfile);
+	mem_fwrite(&fsevh, 1, sizeof(fsevh), outfile);
+	mem_fwrite(fseveh, 1, sizeof(fsb5_sample_extra_vorbis_entry_header) * packet_count, outfile);
 	uint8_t data_padding_arr[32];
 	memset(data_padding_arr, 0, sizeof(data_padding_arr));
-	fwrite(data_padding_arr, 1, data_padding, outfile);
+	mem_fwrite(data_padding_arr, 1, data_padding, outfile);
 	ll_current = ll;
 	while (ll_current != NULL)
 	{
 		uint16_t packet_length;
 		packet_length = ll_current->bytes;
-		fwrite(&packet_length, 1, sizeof(packet_length), outfile);
-		fwrite(ll_current->packet, 1, ll_current->bytes, outfile);
+		mem_fwrite(&packet_length, 1, sizeof(packet_length), outfile);
+		mem_fwrite(ll_current->packet, 1, ll_current->bytes, outfile);
 		ll_current = ll_current->next;
 	}
-	fwrite(data_padding_arr, 1, total_data_padding, outfile);
+	mem_fwrite(data_padding_arr, 1, total_data_padding, outfile);
+
+	kInfoLog += "Done.\n";
+
+	// Copy data into our output data
+	*outSize = outfile.size();
+	*outData = malloc(outfile.size());
+	memcpy(*outData, outfile.data(), outfile.size());
 
-	fprintf(stderr, "Done.\n");
-	return (0);
+	return true;
 }
